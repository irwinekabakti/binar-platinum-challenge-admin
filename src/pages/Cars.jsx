import React, { useState, useEffect } from "react";
import NavbarAdmin from "../components/Navbar/Navbar";
import { Button, Card, Col, Container, Modal, Row } from "react-bootstrap";
import { useNavigate, useSearchParams } from "react-router-dom";
import CarModal from "../assets/car-box.svg";
import moment from "moment";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import {
  faClock,
  faEdit,
  faPlus,
  faTrash,
  faUser,
} from "@fortawesome/free-solid-svg-icons";
import classes from "./Cars.module.css";
import { carsDashboard } from "../store/action/dashboard-slice";
import { useDispatch, useSelector } from "react-redux";
import { deletedCarDashboard } from "../store/action/dashboard-slice";

const Cars = () => {
  const [category, setCategory] = useState(null);
  const [id, setId] = useState(null);
  const [showModal, setShowModal] = useState(false);
  const [showHeaderModal, setShowHeaderModal] = useState(false);
  const [showFormModal, setShowFormModal] = useState(false);
  const navigate = useNavigate();
  const dispatch = useDispatch();
  const selector = useSelector((state) => state.dashboardStore);
  const displayedCars = selector.dataCars;

  let [searchParams, setSearchParams] = useSearchParams();
  const checkFormSuccess = () => {
    searchParams.get("formSuccess") && setShowFormModal(true);
    setSearchParams({});
  };

  const handlingFilterAllCar = () => {
    setCategory(null);
  };

  const handlingFilterSmallCar = () => {
    setCategory("small".toLocaleLowerCase());
  };

  const handlingFilterMediumCar = () => {
    setCategory("medium".toLocaleLowerCase());
  };

  const handlingFilterLargeCar = () => {
    setCategory("large".toLocaleLowerCase());
  };

  const toAddNewCar = () => {
    navigate("/cars/add-new");
  };

  const handleShowFormModal = () => {
    setShowFormModal(false);
  };

  const handleShowHeaderModal = () => {
    setShowHeaderModal(false);
  };

  const handleShowModal = () => {
    setShowModal(false);
  };

  const FormSuccessModal = (props) => {
    return (
      <Modal {...props} size="sm" className="special_modal">
        <Modal.Body
          className={`${classes.bodyModalSuccess} d-flex justify-content-center align-text-center text-white fw-bold`}>
          Data Berhasil Disimpan
        </Modal.Body>
      </Modal>
    );
  };

  const OpenHeaderModal = (props) => {
    return (
      <Modal {...props} size="sm" className="special_modal">
        <Modal.Body
          className={`${classes.bodyModalDelete} d-flex justify-content-center align-text-center text-primary fw-bold`}>
          Data Berhasil di Hapus
        </Modal.Body>
      </Modal>
    );
  };

  useEffect(() => {
    if (id !== null) {
      setShowModal(true);
    }
  }, [id]);

  const deleteCar = async () => {
    try {
      dispatch(deletedCarDashboard(id))
        .unwrap()
        .then(() => {
          getCars();
          // navigate("/cars");
        });
    } catch (error) {
      // console.log(error);
      alert(error);
    }
  };

  const getCars = () => {
    dispatch(carsDashboard())
      .then(() => {
        checkFormSuccess();
      })
      .catch((error) => {
        alert(error);
      });
  };

  const isModalOpen = () => {
    deleteCar();
    setShowModal(false);
    setShowHeaderModal(true);
  };

  const OpenModal = (props) => {
    return (
      <Modal
        size="sm"
        className="w-10"
        aria-labelledby="container-modal-title-vcenter"
        centered
        {...props}
        backdrop="static"
        keyboard={false}>
        <Modal.Body>
          <div className="w-100 d-flex justify-content-center mt-4">
            <img src={CarModal} alt="mobil-infoModal" />
          </div>
          <div className="w-100 d-flex justify-content-center mt-3">
            <p>
              Setelah dihapus data mobil tidak dapat dikembalikan. Yakin ingin
              menghapus?
            </p>
          </div>

          <div className="container px-4 mt-5">
            <div className="row gx-5">
              <Col>
                <Button
                  style={{ width: "100%" }}
                  variant="primary"
                  onClick={isModalOpen}>
                  Ya
                </Button>
              </Col>
              <Col>
                <Button
                  style={{ width: "100%" }}
                  variant="outline-primary"
                  onClick={props.onHide}>
                  Tidak
                </Button>
              </Col>
            </div>
          </div>
        </Modal.Body>
      </Modal>
    );
  };

  useEffect(() => {
    getCars();
  }, []);

  const filterCars = () => {
    let renderCars = displayedCars;

    if (category) {
      renderCars = renderCars.filter((car) => {
        if (car.category === category) return car;
      });
    }

    return (
      <Row className="g-4 mb-5 mt-0">
        {renderCars.map((car) => (
          <Col
            xs={12}
            md={6}
            lg={4}
            xxl={4}
            className="containerCardCars"
            key={car.id}>
            <Card className="cardCars mt-0">
              <div className={classes.wrapperImageCars}>
                <Card.Img src={car.image} />
              </div>
              <Card.Body>
                <Card.Text className="fw-bold">{car.name}</Card.Text>
                <Card.Text className="fw-bold">
                  Rp {car.price.toLocaleString("id-ID")} / Hari
                </Card.Text>
                {car.category.toLowerCase() === "small" ? (
                  <p>
                    <small>
                      <FontAwesomeIcon icon={faUser} /> 2 - 4 People
                    </small>
                  </p>
                ) : null}
                {car.category.toLowerCase() === "medium" ? (
                  <p>
                    <small>
                      <FontAwesomeIcon icon={faUser} /> 4 - 6 People
                    </small>
                  </p>
                ) : null}
                {car.category.toLowerCase() === "large" ? (
                  <p>
                    <small>
                      <FontAwesomeIcon icon={faUser} /> 6 - 8 People
                    </small>
                  </p>
                ) : null}

                <p>
                  <small>
                    <FontAwesomeIcon icon={faClock} /> Updated at{" "}
                    {moment(car.updatedAt).format("DD MMMM YYYY h:mm A")}
                  </small>
                </p>
                <div className="d-flex justify-content-around">
                  <Button
                    variant="outline-danger"
                    size="lg"
                    className={`d-flex align-items-center ${classes.btnDelete}`}
                    style={{ width: "fit-content" }}
                    onClick={() => {
                      setId(car.id);
                    }}>
                    <FontAwesomeIcon icon={faTrash} className="me-2 " />
                    Delete
                  </Button>
                  <Button
                    variant="outline-none"
                    size="lg"
                    className={`d-flex align-items-center text-white ${classes.btnEdit}`}
                    style={{ width: "fit-content" }}
                    onClick={() => navigate(`/cars/edit/${car.id}`)}>
                    <FontAwesomeIcon icon={faEdit} className="me-2" />
                    Edit
                  </Button>
                </div>
              </Card.Body>
            </Card>
          </Col>
        ))}
      </Row>
    );
  };

  return (
    <Container fluid className={`p-0 m-0 ${classes.containerCarsPage}`}>
      <NavbarAdmin currentPage="cars" />
      <Row className="m-0">
        <Col
          xs="auto"
          className={`d-none d-md-block h-100 ${classes.colCarsPage}`}></Col>
        <Col>
          <p className="mt-4">
            <strong>Cars {">"}</strong> List Car
          </p>
          <h5 className="mt-4">
            <strong>List Car</strong>
          </h5>
          <div className="d-flex justify-content-end">
            <Button
              className={`pe-3 mb-2 fw-bold ${classes.btnAddCar}`}
              onClick={toAddNewCar}>
              <FontAwesomeIcon icon={faPlus} className="me-2 " />
              Add New Car
            </Button>
          </div>
          <Button
            className={`me-3 mb-2 ${classes.btnFilterCars}`}
            onClick={handlingFilterAllCar}>
            All
          </Button>
          <Button
            className={`me-3 mb-2 ${classes.btnFilterCars}`}
            onClick={handlingFilterSmallCar}>
            2 - 4 people
          </Button>
          <Button
            className={`me-3 mb-2 ${classes.btnFilterCars}`}
            onClick={handlingFilterMediumCar}>
            4 - 6 people
          </Button>
          <Button
            className={`me-3 mb-2 ${classes.btnFilterCars}`}
            onClick={handlingFilterLargeCar}>
            6 - 8 people
          </Button>
          {filterCars()}
        </Col>
      </Row>
      <FormSuccessModal show={showFormModal} onHide={handleShowFormModal} />
      <OpenHeaderModal show={showHeaderModal} onHide={handleShowHeaderModal} />
      <OpenModal show={showModal} onHide={handleShowModal} />
    </Container>
  );
};

export default Cars;
