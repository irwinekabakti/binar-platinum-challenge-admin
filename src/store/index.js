import { configureStore } from "@reduxjs/toolkit";

import authAdminSlice from "./action/admin-slice";
import dashboardSlice from "./action/dashboard-slice";

export const store = configureStore({
  reducer: {
    adminStore: authAdminSlice.reducer,
    dashboardStore: dashboardSlice.reducer,
  },
});
