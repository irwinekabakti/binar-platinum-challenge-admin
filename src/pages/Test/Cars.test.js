import { describe, test, expect } from "@jest/globals";
import Renderer from "react-test-renderer";
import { render, screen, waitFor } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import { Provider } from "react-redux";
// import { renderWithProviders } from "../../utils/test-utils";
import { store } from "../../store";
import Cars from "../Cars";

describe('UI Cars Test', () => {
  test('UI Cars should match snapshot', () => {
    const snapshot = Renderer.create(
      <Provider store={store}>
        <BrowserRouter>
          <Cars />
        </BrowserRouter>
      </Provider>
    ).toJSON()

    expect(snapshot).toMatchSnapshot()
  })
})

describe('Test component Cars', () => {
  test('it should be render component Cars', () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <Cars />
        </BrowserRouter>
      </Provider>
    )

    const containerTitle = screen.getByTestId('container-Title')
    const titleCars = screen.getByTestId('title-Cars')
    const titleCars2 = screen.getByTestId('title-Cars-2')

    expect(containerTitle).toContainElement(titleCars)
    expect(containerTitle).toContainElement(titleCars2)

  })
})

