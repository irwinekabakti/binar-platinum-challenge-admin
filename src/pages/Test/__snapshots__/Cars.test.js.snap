// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI Cars Test UI Cars should match snapshot 1`] = `
<div
  className="p-0 m-0 containerCarsPage container-fluid"
>
  <div
    className="bg-white navigation-panel position-absolute h-100  navPanelClass"
  >
    <div
      className="cars-menu w-100 flex-column"
    >
      <p
        className="w-100 ps-4 pe-3 titleCars"
      >
        CARS
      </p>
      <ul
        className="p-0 listCars"
      >
        <li
          className="ps-4 pe-3 listNameCars"
        >
          List Car
        </li>
      </ul>
    </div>
  </div>
  <div
    className="navigation-menu position-absolute flex-column  navMenuClass  nav"
    onKeyDown={[Function]}
  >
    <a
      className="navigationTab w-100 d-flex flex-column justify-content-center align-items-center text-center false nav-link"
      data-rr-ui-event-key={null}
      href="#"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      style={
        Object {
          "height": "64px",
        }
      }
      tabIndex={0}
    >
      <img
        alt="dashboard"
        src="fi_home.svg"
      />
      <p
        className="m-0 text-white navTitle"
      >
        Dashboard
      </p>
    </a>
    <a
      className="navigation-tab w-100 d-flex flex-column justify-content-center align-items-center text-center selected nav-link"
      data-rr-ui-event-key={null}
      href="#"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      style={
        Object {
          "height": "64px",
        }
      }
      tabIndex={0}
    >
      <img
        alt="cars"
        src="fi_truck.svg"
      />
      <p
        className="m-0 text-white navTitle"
      >
        Cars
      </p>
    </a>
  </div>
  <nav
    className="bg-white p-0 mainNavbar navbar navbar-expand navbar-light"
  >
    <div
      className="h-100 p-0 d-flex justify-content-between align-items-center container-fluid"
    >
      <div
        className="h-100 d-flex colNav col-auto"
      >
        <button
          aria-label="Navigation Menu"
          className="menu-button d-flex justify-content-center align-items-center position-relative btnNavSideBar btn btn-primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <div
            className="iconSideNav"
          />
        </button>
        <div
          className="h-100 w-100 d-flex align-items-center px-4 spaceNavTop"
        >
          <div
            className="iconTopNav"
          />
        </div>
      </div>
      <div
        className="h-100 d-flex justify-content-between col"
      >
        <div
          className="h-100 w-100 d-flex justify-content-center align-items-center containerBurgerButton"
          data-testid="container-burger-button-custom"
        >
          <button
            className="bg-white p-1 d-flex flex-column border-0 btnBurgerButton btn btn-primary"
            data-testid="burger-button-custom"
            disabled={false}
            type="button"
          >
            <span
              className="bg-dark burgerButton"
            />
            <span
              className="bg-dark burgerButton"
            />
            <span
              className="bg-dark burgerButton"
            />
          </button>
        </div>
        <div
          className="d-flex h-100"
        >
          <form
            className="d-none d-md-flex align-items-center pe-3"
          >
            <fieldset
              className="position-relative d-flex"
            >
              <input
                className="formNavbar formNavbar form-control"
                placeholder="Search"
                type="text"
              />
              <img
                alt="search"
                aria-hidden="true"
                className="iconSearch"
                src="fi_search.svg"
              />
              <button
                className="fw-bold btnNavbarSearch btn btn-primary"
                disabled={false}
                type="button"
              >
                Search
              </button>
            </fieldset>
          </form>
          <div
            className="h-100 w-100 pe-3 d-flex justify-content-center align-items-center containerProfile"
          >
            <div
              className="dropdown nav-item"
            >
              <button
                aria-expanded={false}
                className="bg-white border-0 d-flex align-items-center admin-menu dropdown-toggle btn btn-primary"
                disabled={false}
                id="dropdown-1"
                onClick={[Function]}
                type="button"
              >
                <div
                  className="rounded-circle d-flex justify-content-center align-items-center ImageProfileNavbar"
                >
                  R
                </div>
                <div
                  className="px-2 text-dark profileNavbar"
                >
                  Marcus Raspot
                </div>
                <img
                  alt="icon-navbar"
                  aria-hidden="true"
                  src="chevron-down.svg"
                />
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </nav>
  <div
    className="m-0 row"
  >
    <div
      className="d-none d-md-block h-100 colCarsPage col-auto"
    />
    <div
      className="col"
      data-testid="container-Title"
    >
      <p
        className="mt-4"
        data-testid="title-Cars"
      >
        <strong>
          Cars 
          &gt;
        </strong>
         List Car
      </p>
      <h5
        className="mt-4"
        data-testid="title-Cars-2"
      >
        <strong>
          List Car
        </strong>
      </h5>
      <div
        className="d-flex justify-content-end"
      >
        <button
          className="pe-3 mb-2 fw-bold btnAddCar btn btn-primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-plus me-2 "
            data-icon="plus"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M240 80c0-17.7-14.3-32-32-32s-32 14.3-32 32V224H32c-17.7 0-32 14.3-32 32s14.3 32 32 32H176V432c0 17.7 14.3 32 32 32s32-14.3 32-32V288H384c17.7 0 32-14.3 32-32s-14.3-32-32-32H240V80z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
          Add New Car
        </button>
      </div>
      <button
        className="me-3 mb-2 btnFilterCars btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        All
      </button>
      <button
        className="me-3 mb-2 btnFilterCars btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        2 - 4 people
      </button>
      <button
        className="me-3 mb-2 btnFilterCars btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        4 - 6 people
      </button>
      <button
        className="me-3 mb-2 btnFilterCars btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        6 - 8 people
      </button>
      <div
        className="g-4 mb-5 mt-0 row"
      />
    </div>
  </div>
</div>
`;
